### PR TITLE
Add support for type 'string' cast operator

### PR DESCRIPTION
**Pull request by smakalias**
## Summary of the changes / Why this improves CrateDB

This PR is to add support the following cast syntax:
`type 'string'` in addition to existing casting alternatives:
- `'string'::type`
- `CAST ( 'string' AS type )`

## Checklist

 - [x] User relevant changes are recorded in ``CHANGES.txt``
 - [x] Touched code is covered by tests
 - [x] Documentation has been updated if necessary
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java [hunk 1]
```diff
@@ -401,13 +396,6 @@ public void testTrimFunctionMissingFromWhenCharsToTrimIsPresentThrowsException()
         assertInstanceOf("TRIM(' ' chars)", FunctionCall.class);
     }
 
-    @Test
-    public void testTrimFunctionInvalidTrimModeThrowsException() {
-        expectedException.expect(ParsingException.class);
-        expectedException.expectMessage("line 1:12: no viable alternative at input 'TRIM(BOTHH ' '");
-        assertInstanceOf("TRIM(BOTHH ' ' FROM chars)", FunctionCall.class);
-    }
```
Comment (smakalias) [inline sql-parser/src/test/java/io/crate/sql/parser/TestSqlParser.java]: Removed as it's a no longer an invalid test. 

`BOTHH ' '` results in a cast expression which will later fail (`Cannot find data type: bothh`).
However, for this context, it is valid - thus It's removed completely.

File: sql-parser/src/main/antlr/SqlBase.g4 [hunk 2]
```diff
@@ -329,7 +330,6 @@ setCmpQuantifier
 datetimeLiteral
```
Comment (kovrus) [inline sql-parser/src/main/antlr/SqlBase.g4]: we can probably drop the `datetimeLiteral` rule?

File: sql/src/test/java/io/crate/integrationtests/PostgresITest.java [hunk 3]
```diff
@@ -601,6 +602,7 @@ public void testStatementThatResultsInParseError() throws Exception {
             stmt.executeQuery();
         }
     }
+    */
```
Comment (smakalias) [inline sql/src/test/java/io/crate/integrationtests/PostgresITest.java]: I'd like to discuss this one as the error seems a bit odd in this case.

This test is now failing with Error `Cannot find data type: fro` instead of `mismatched input 'sys'`

I've tried a similar query in pg:
```
mydb=> select name fro test1;
2019-04-15 12:51:11.551 EEST [65047] ERROR:  syntax error at or near "test1" at character 17
2019-04-15 12:51:11.551 EEST [65047] STATEMENT:  select name fro test1;
ERROR:  syntax error at or near "test1"
LINE 1: select name fro test1;
```
It's not as clear I would have expected but it'd better than the type error.
So it seems that the error is handled differently.

What's your opinion on that?

**Comment by smakalias** [comment]
note to self: fix typo in commit message

File: sql/src/test/java/io/crate/integrationtests/PostgresITest.java (hunk 3 above)
Comment (kovrus) [reply sql/src/test/java/io/crate/integrationtests/PostgresITest.java]: hmm that seems to behave really wrong, it should not try to resolve a data type. It seems that this parser rule is not even matched in this case
```
selectItem:
   expr (AS? ident)?  
```

Comment (smakalias) [reply sql/src/test/java/io/crate/integrationtests/PostgresITest.java]: hmm. I think I should really make this rule strict compared to the other notations (`cast()`, `::`)

**Comment by kovrus** [comment]
just another thing, the story does not require implementing this kind of cast, it is probably enough to properly handle the supported by parser datetime literal.

File: sql/src/test/java/io/crate/integrationtests/PostgresITest.java (hunk 3 above)
Comment (smakalias) [reply sql/src/test/java/io/crate/integrationtests/PostgresITest.java]: @kovrus `selectItem` rule is matched, the issue is that it matches the `alternateCast` rule later on.

Comment (kovrus) [reply sql/src/test/java/io/crate/integrationtests/PostgresITest.java]: yep, right

Comment (smakalias) [reply sql/src/test/java/io/crate/integrationtests/PostgresITest.java]: @kovrus I finally restricted the rule to accept a stringLiteral which is actually the use case anyway

File: sql-parser/src/main/antlr/SqlBase.g4 (hunk 2 above)
Comment (smakalias) [reply sql-parser/src/main/antlr/SqlBase.g4]: yeah, I guess we could drop the whole thing?

@mfussenegger any reason to keep this?

File: sql-parser/src/main/antlr/SqlBase.g4 [hunk 4]
```diff
@@ -215,6 +215,7 @@ valueExpression
     | left=valueExpression operator=(PLUS | MINUS) right=valueExpression             #arithmeticBinary
     | left=valueExpression CONCAT right=valueExpression                              #concatenation
     | valueExpression CAST_OPERATOR dataType                                         #doubleColonCast
+    | dataType stringLiteral                                                         #fromStringLiteralCast
```
Comment (mfussenegger) [inline sql-parser/src/main/antlr/SqlBase.g4]: Should we restrict this to simple types? (exclude arrays, object definition).

E.g. from the postgres docs:

> To avoid syntactic ambiguity, the type 'string' syntax can only be used to specify the type of a simple literal constant. Another restriction on the type 'string' syntax is that it does not work for array types; use :: or CAST() to specify the type of an array constant.

SQL Standard also doesn't require us to allow it for all types.

File: sql-parser/src/main/antlr/SqlBase.g4 (hunk 2 above)
Comment (mfussenegger) [reply sql-parser/src/main/antlr/SqlBase.g4]: No, if it doesn't break anything?

File: blackbox/docs/appendices/release-notes/unreleased.rst [hunk 5]
```diff
@@ -144,6 +144,10 @@ Deprecations
 Changes
 =======
 
+- Added support for the :ref:`type 'string' <type_cast_from_string_literal>`
+  cast operator, which is used for the initial assignment of a type to
```
Comment (mfussenegger) [inline blackbox/docs/appendices/release-notes/unreleased.rst]: This sounds somehow very sophisticated - can we explain this in simpler terms?

File: blackbox/docs/general/ddl/data-types.rst [hunk 6]
```diff
@@ -854,6 +854,37 @@ Trying to cast a ``text`` to ``integer``, will fail with ``cast`` if
     +---------------------------+
     SELECT 1 row in set (... sec)
 
+.. _type_cast_from_string_literal:
+
+``type 'string'``
+-----------------
+
+This cast operation is applied to a string literal and it is effectively
```
Comment (mfussenegger) [inline blackbox/docs/general/ddl/data-types.rst]: Same here

File: sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java [hunk 7]
```diff
@@ -1377,7 +1375,16 @@ public Node visitDoubleColonCast(SqlBaseParser.DoubleColonCastContext context) {
 
     @Override
     public Node visitFromStringLiteralCast(SqlBaseParser.FromStringLiteralCastContext context) {
-        return new Cast((Expression) visit(context.stringLiteral()), (ColumnType) visit(context.dataType()));
+        ColumnType targetType = (ColumnType) visit(context.dataType());
+        if (targetType.type() != ColumnType.Type.PRIMITIVE) {
+            throw new UnsupportedOperationException("type 'string' cast notation only supports primitive types. " +
+                                                    "Use '::' or cast() operator instead.");
+        }
+        if (targetType instanceof ObjectColumnType) {
```
Comment (kovrus) [inline sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: wouldn't it be enough just to the date type check that we have above?

Comment (smakalias) [reply sql-parser/src/main/java/io/crate/sql/parser/AstBuilder.java]: no, for some reason the `ObjectColumnType` results in having type: `Primitive` which is a bit weird. I'll try to change this if there's no big impact.

Thanks for catching this.
